flocker/node/test/test_script.py

```diff
@@ -303,6 +307,49 @@ def test_logging(self, log_name):
         self.assertIn(log_message, logfile.getContent())
 
 
+def mimic_for_test(test_case):
+    """
+    Start a mimic server in the background on an ephemeral port and return the
+    port number.
+
+    This is used in synchronous test cases so I can't launch the mimic service
+    in process.
+
+    Parsing the logs for the chosen port number is ugly, but ``find_free_port``
+    kept returning ports that were in use when mimic attempted to bind to them.
+    """
+    log = test_case.make_temporary_path()
+    stdout = test_case.make_temporary_path()
+    stderr = test_case.make_temporary_path()
+    with stdout.open('w') as stdout, stderr.open('w') as stderr:
+        p = Popen(['twistd', '--nodaemon', '--logfile', log.path,
+                   'mimic', '--listen', '0', '--realtime'],
+                  stdin=open(os.devnull), stdout=stdout, stderr=stderr,
+                  close_fds=True)
+
+    def cleanup():
+        p.terminate()
+        p.wait()
+    test_case.addCleanup(cleanup)
+
+    poll_until(
+        predicate=log.exists,
+        steps=repeat(1, 5)
+    )
+
+    def port_from_log():
+        for line in log.open():
+            match = re.search(r"Site starting on (\d+)$", line)
```

Comment:
Might be an irrelevant tangent, but you can probably tell the child to listen on an inherited fd using, eg, the systemd endpoint (writing that, now seems silly to only have support for this via "systemd" instead of a generic inherited-fd endpoint, oh well).  It's probably a little fiddling around to set things up as systemd would have (and thus make the systemd endpoint actually work) but it means not having to parse the logs to discover the port.